### ISSUE TITLE
NNBD: Analyzer throws error for `Set<double> x2 = {1 + 1};` where dart passes here

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#43139]
Dart SDK version: 2.10.0-27.0.dev (dev) (Thu Aug 13 23:48:41 2020 -0700) on "windows_x64"

The following source code example passes with Dart without errors whereas analyzer throws a compile time error for `Set<double> x2 = {1 + 1};` and passes with  `Set<double> x1 = {1};` if null safety is turned on:
``` dart
main() {
  Set<double> x1 = {1};
  Set<double> x2 = {1 + 1};
}
```

Both analyzer and dart throws a compile error with nnbd turned off here.

I am not quite sure which behavior is correct, but anyway all the tools should work in the same manner.

Sample output is:

> $> dart test.dart
> test.dart:3:23: Error: A value of type 'int' can't be assigned to a variable of type 'double'.
>   Set<double> x2 = {1 + 1};
>                       ^
> $> dartanalyzer test.dart
> Analyzing test.dart...
>   error - The element type 'int' can't be assigned to the set type 'double'. - test.dart:3:21 - set_element_type_not_assignable
>   hint - The value of the local variable 'x1' isn't used. - test.dart:2:15 - unused_local_variable
>   hint - The value of the local variable 'x2' isn't used. - test.dart:3:15 - unused_local_variable
> 1 error and 2 hints found.
> 
> $> dart --enable-experiment=non-nullable test.dart
> 
> $> dartanalyzer --enable-experiment=non-nullable test.dart
> Analyzing test.dart...
>   error - The element type 'int' can't be assigned to the set type 'double'. - test.dart:3:21 - set_element_type_not_assignable
>   hint - The value of the local variable 'x1' isn't used. - test.dart:2:15 - unused_local_variable
>   hint - The value of the local variable 'x2' isn't used. - test.dart:3:15 - unused_local_variable
> 1 error and 2 hints found.
> 
> $>

**Comment by lrhn** [comment]
Dart promotes integer literals to double literals in a double context. That''s what `double x = 1;` works, and also why `Set<double> s = {1};` works.

However, there is traditionally no `double` context for the integers in `double x = 1 + 1;`. The former `1` is a receiver for the `+` method, the object it's being called on, and it has no context type at all. The latter `1` is an argument to the `num operator+(num)` operator of `int`, so its context type is `num`. The result is still an `int` because of the special rules for integer arithmetic.

With null safety, we are currently in the process of *changing* those rules to be more clever ([desing doc](https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/number-operation-typing.md)). That's necessary because null safe Dart does not have implicit downcasts, so ending up with a `num` result in cases where we could derive that the result is `double`  is introducing new annoying errors.

One of these changes is that doing `1 + x` in a double context will figure out that the only way `int + something` is going to be a `double` is if the something is a double, so it will give `x` a context type of `double`. Then `double x = 1 + 1;` will convert the latter `1` to a double literal, and the result becomes a `double`.
What you are seeing is likely that the front-end has made this change, but the analyzer hasn't made it yet. It will be implemented on both engines before null-safety is released.

**Comment by stereotype441** [comment]
Fixed in 9af30dbb06d164c926e465137e179c057059b259
